echidna/core/smear.py

```diff
@@ -0,0 +1,224 @@
+
+class Smear(object):
+    """ This class smears the energy and radius of a spectra.
+
+    The class can recieve energy and radius as individual data points or a 
+    1 dimensional numpy array to smear which is then returned. 2d and 3d 
+    arrays with linked energy, radius and time information is yet to be
+    implemented.
+
+    Attributes:
+      _light_yield (float): Number of PMT hits expected for a MeV energy deposit in NHit/MeV
+      _position_resolution (float): Sigma in mm
+    """
+    _light_yield = 200. # NHit per MeV
+    _position_resolution = 100. # mm
+
+    def __init__(self):
+        """ Initialise the Smear class by seeding the random number generator
+        """
+        numpy.random.seed()
+
+    def bin_1d_array(self, array, bins):
+        """ Sorts a 1 dimensional array and bins it
+        
+        Args:
+          array (numpy array): To sort and bin
+          bins (list): Upper limit of bins
+
+        Returns:
+          A 1 dimensional numpy array, sorted and binned.
+        """
+        array = numpy.sort(array)
+        split_at = array.searchsorted(bins)
+        return numpy.split(array,split_at)
+        
+    def floor_to_bin(self, x,bin_size):
+        """ Rounds down value bin content to lower edge of nearest bin.
+        
+        Args:
+          x (float): Value to round down
+          bin_size (float): Width of a bin
+
+        Returns:
+          Value of nearest lower bin edge
+        """
+        dp = abs(decimal.Decimal(str(bin_size)).as_tuple().exponent)
+        coef = numpy.power(10,dp)
+        return numpy.floor(coef*(x//bin_size)*bin_size)/coef
+
+    def get_energy_sigma(self,energy):
+        """ Calculates sigma at a given energy.
+        
+        Args:
+          energy (float): Energy value of data point(s)
+        
+        Returns:
+          Sigma equivalent to sqrt(energy/_light_yield)
+        """
+        return numpy.sqrt(energy/self._light_yield)
+
+    def smear_energy_0d(self,energy):
+        """ Smears a single energy value
+        
+        Args:
+          energy (float): Value to smear
+
+        Returns:
+          Smeared energy value
+        """
+        sigma = self.get_energy_sigma(energy)
+        return numpy.fabs(numpy.random.normal(energy,sigma))   
+
+    def smear_energy_1d(self,energies, bins, binned = False):
+        """ Smears a 1 dimensional array of energy values
+        
+        Args:
+          energies (numpy array): Values to smear
+          bins (list): Upper edge of bins for array
+          binned (bool): Is the array already binned? (True or False)
+
+        Returns:
+          Smeared and sorted 1 dimensional numpy array of energy values
+        """
+        if binned == False:
+            energies = self.bin_1d_array(energies,bins)
+        bin_size = bins[1]-bins[0]
+        smeared_energies = []
+        for entry in energies:
+            if not entry.any():
+                # Bin is empty
+                continue
```

Comment:
Changed it to the former. An error message isn't neccessary for this as it is likely that a PDF will have bins with no entries in it.